subprojects/execution/src/main/java/org/gradle/internal/execution/impl/steps/ExecuteStep.java

```diff
@@ -45,7 +46,11 @@ public ExecuteStep(
     public Result execute(Context context) {
         UnitOfWork work = context.getWork();
 
-        outputChangeListener.beforeOutputChange();
+        Optional<? extends Iterable<String>> changingOutputs = work.getChangingOutputs();
+        changingOutputs.ifPresent(outputs -> outputChangeListener.beforeOutputChange(outputs));
+        if (!changingOutputs.isPresent()) {
+            outputChangeListener.beforeOutputChange();
+        }
```

Comment:
Why do we need to do it like this?

Reply:
The problem I am trying to solve here is the following: I don't want to invalidate the whole file system mirror for artifact transformations, since I know exactly which parts need to be invalidated. For tasks though, we still need to invalidate everything.

So the `UnitOfWork` needs to provide a way to say that either
- it knows exactly which parts of the file system mirror to invalidate and provides it to the executor (for transformations)
- it cannot be sure which parts of the file system mirror need to be invalidated, so the whole file system mirror needs to be invalidated (tasks).

Reply:
Added the Javadoc.